In [2]:
import numpy as np

In [3]:
import pandas as pd

In [4]:
df=pd.read_csv(r"C:\Users\DELL\Desktop\Sipalaya Data Science\datas\smartphonesss.csv")
df

Unnamed: 0                 model      price  rating  \
0          917      Nokia 150 (2020)     ₹2,499     NaN   
1          462   Xiaomi Mi Mix Alpha  ₹1,99,990     NaN   
2            3  Motorola Moto G62 5G    ₹14,999    81.0   
3          739             Realme Q3    ₹14,999    80.0   
4          117        iQOO Z6 Pro 5G    ₹21,788    78.0   
..         ...                   ...        ...     ...   
95         682          LG V60 ThinQ    ₹79,990     NaN   
96         350     Motorola Moto G22     ₹9,499    75.0   
97         785    Xiaomi 12 Ultra 5G    ₹75,999     NaN   
98         594         iQOO 9 Pro 5G    ₹59,990    88.0   
99         297         OPPO F21s Pro    ₹25,895    82.0   

                                                  sim  \
0                                            Dual Sim   
1                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
2                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
3                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
4                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
..                                                ...   
95            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
96                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
97  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
98            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
99                 Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   

                                           processor  \
0                                            No Wifi   
1     Snapdragon 855+, Octa Core, 2.96 GHz Processor   
2      Snapdragon  695, Octa Core, 2.2 GHz Processor   
3    Snapdragon  750G , Octa Core, 2.2 GHz Processor   
4     Snapdragon  778G, Octa Core, 2.4 GHz Processor   
..                                               ...   
95    Snapdragon  865, Octa Core, 2.84 GHz Processor   
96          Helio  G37, Octa Core, 2.3 GHz Processor   
97  Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor   
98     Snapdragon 8 Gen1, Octa Core, 3 GHz Processor   
99     Snapdragon  695, Octa Core, 2.2 GHz Processor   

                          ram                                   battery  \
0                4 MB inbuilt                          1020 mAh Battery   
1   12 GB RAM, 512 GB inbuilt   4050 mAh Battery with 40W Fast Charging   
2    6 GB RAM, 128 GB inbuilt       5000 mAh Battery with Fast Charging   
3    6 GB RAM, 128 GB inbuilt   5000 mAh Battery with 30W Fast Charging   
4    6 GB RAM, 128 GB inbuilt   4700 mAh Battery with 66W Fast Charging   
..                        ...                                       ...   
95   8 GB RAM, 128 GB inbuilt       5000 mAh Battery with Fast Charging   
96    4 GB RAM, 64 GB inbuilt   5000 mAh Battery with 20W Fast Charging   
97  12 GB RAM, 256 GB inbuilt   4800 mAh Battery with 67W Fast Charging   
98   8 GB RAM, 256 GB inbuilt  4700 mAh Battery with 120W Fast Charging   
99   8 GB RAM, 128 GB inbuilt   4500 mAh Battery with 33W Fast Charging   

                                              display  \
0                    2.4 inches, 240 x 320 px Display   
1                 7.92 inches, 2088 x 2250 px Display   
2   6.55 inches, 1080 x 2400 px, 120 Hz Display wi...   
3   6.5 inches, 1080 x 2400 px, 120 Hz Display wit...   
4   6.44 inches, 1080 x 2404 px, 90 Hz Display wit...   
..                                                ...   
95  6.8 inches, 1080 x 2460 px Display with Punch ...   
96  6.53 inches, 720 x 1600 px, 90 Hz Display with...   
97  6.73 inches, 1440 x 3200 px, 120 Hz Display wi...   
98  6.78 inches, 1440 x 3200 px, 120 Hz Display wi...   
99  6.43 inches, 1080 x 2400 px Display with Punch...   

                                               camera  \
0                                  0.3 MP Rear Camera   
1           108 MP + 20 MP + 12 MP Triple Rear Camera   
2   50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
3   48 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
4   64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
..       

### 1.Consider the Dataset Smartphone contains Price columns price - has unneccesary '₹' and ',' replace the unnecessary character and convert into numeric columns

In [10]:
df['price']=df['price'].str.replace('₹','').str.replace(',','')

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  100 non-null    int64  
 1   model       100 non-null    object 
 2   price       100 non-null    object 
 3   rating      78 non-null     float64
 4   sim         100 non-null    object 
 5   processor   100 non-null    object 
 6   ram         100 non-null    object 
 7   battery     100 non-null    object 
 8   display     100 non-null    object 
 9   camera      100 non-null    object 
 10  card        99 non-null     object 
 11  os          99 non-null     object 
dtypes: float64(1), int64(1), object(10)
memory usage: 9.5+ KB


In [12]:
df

Unnamed: 0                 model   price  rating  \
0          917      Nokia 150 (2020)    2499     NaN   
1          462   Xiaomi Mi Mix Alpha  199990     NaN   
2            3  Motorola Moto G62 5G   14999    81.0   
3          739             Realme Q3   14999    80.0   
4          117        iQOO Z6 Pro 5G   21788    78.0   
..         ...                   ...     ...     ...   
95         682          LG V60 ThinQ   79990     NaN   
96         350     Motorola Moto G22    9499    75.0   
97         785    Xiaomi 12 Ultra 5G   75999     NaN   
98         594         iQOO 9 Pro 5G   59990    88.0   
99         297         OPPO F21s Pro   25895    82.0   

                                                  sim  \
0                                            Dual Sim   
1                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
2                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
3                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
4                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
..                                                ...   
95            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
96                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
97  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
98            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
99                 Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   

                                           processor  \
0                                            No Wifi   
1     Snapdragon 855+, Octa Core, 2.96 GHz Processor   
2      Snapdragon  695, Octa Core, 2.2 GHz Processor   
3    Snapdragon  750G , Octa Core, 2.2 GHz Processor   
4     Snapdragon  778G, Octa Core, 2.4 GHz Processor   
..                                               ...   
95    Snapdragon  865, Octa Core, 2.84 GHz Processor   
96          Helio  G37, Octa Core, 2.3 GHz Processor   
97  Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor   
98     Snapdragon 8 Gen1, Octa Core, 3 GHz Processor   
99     Snapdragon  695, Octa Core, 2.2 GHz Processor   

                          ram                                   battery  \
0                4 MB inbuilt                          1020 mAh Battery   
1   12 GB RAM, 512 GB inbuilt   4050 mAh Battery with 40W Fast Charging   
2    6 GB RAM, 128 GB inbuilt       5000 mAh Battery with Fast Charging   
3    6 GB RAM, 128 GB inbuilt   5000 mAh Battery with 30W Fast Charging   
4    6 GB RAM, 128 GB inbuilt   4700 mAh Battery with 66W Fast Charging   
..                        ...                                       ...   
95   8 GB RAM, 128 GB inbuilt       5000 mAh Battery with Fast Charging   
96    4 GB RAM, 64 GB inbuilt   5000 mAh Battery with 20W Fast Charging   
97  12 GB RAM, 256 GB inbuilt   4800 mAh Battery with 67W Fast Charging   
98   8 GB RAM, 256 GB inbuilt  4700 mAh Battery with 120W Fast Charging   
99   8 GB RAM, 128 GB inbuilt   4500 mAh Battery with 33W Fast Charging   

                                              display  \
0                    2.4 inches, 240 x 320 px Display   
1                 7.92 inches, 2088 x 2250 px Display   
2   6.55 inches, 1080 x 2400 px, 120 Hz Display wi...   
3   6.5 inches, 1080 x 2400 px, 120 Hz Display wit...   
4   6.44 inches, 1080 x 2404 px, 90 Hz Display wit...   
..                                                ...   
95  6.8 inches, 1080 x 2460 px Display with Punch ...   
96  6.53 inches, 720 x 1600 px, 90 Hz Display with...   
97  6.73 inches, 1440 x 3200 px, 120 Hz Display wi...   
98  6.78 inches, 1440 x 3200 px, 120 Hz Display wi...   
99  6.43 inches, 1080 x 2400 px Display with Punch...   

                                               camera  \
0                                  0.3 MP Rear Camera   
1           108 MP + 20 MP + 12 MP Triple Rear Camera   
2   50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
3   48 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
4   64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
..                                           

In [16]:
df['price']=df['price'].astype(int)

In [17]:
df

Unnamed: 0                 model   price  rating  \
0          917      Nokia 150 (2020)    2499     NaN   
1          462   Xiaomi Mi Mix Alpha  199990     NaN   
2            3  Motorola Moto G62 5G   14999    81.0   
3          739             Realme Q3   14999    80.0   
4          117        iQOO Z6 Pro 5G   21788    78.0   
..         ...                   ...     ...     ...   
95         682          LG V60 ThinQ   79990     NaN   
96         350     Motorola Moto G22    9499    75.0   
97         785    Xiaomi 12 Ultra 5G   75999     NaN   
98         594         iQOO 9 Pro 5G   59990    88.0   
99         297         OPPO F21s Pro   25895    82.0   

                                                  sim  \
0                                            Dual Sim   
1                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
2                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
3                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
4                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
..                                                ...   
95            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
96                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
97  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
98            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
99                 Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   

                                           processor  \
0                                            No Wifi   
1     Snapdragon 855+, Octa Core, 2.96 GHz Processor   
2      Snapdragon  695, Octa Core, 2.2 GHz Processor   
3    Snapdragon  750G , Octa Core, 2.2 GHz Processor   
4     Snapdragon  778G, Octa Core, 2.4 GHz Processor   
..                                               ...   
95    Snapdragon  865, Octa Core, 2.84 GHz Processor   
96          Helio  G37, Octa Core, 2.3 GHz Processor   
97  Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor   
98     Snapdragon 8 Gen1, Octa Core, 3 GHz Processor   
99     Snapdragon  695, Octa Core, 2.2 GHz Processor   

                          ram                                   battery  \
0                4 MB inbuilt                          1020 mAh Battery   
1   12 GB RAM, 512 GB inbuilt   4050 mAh Battery with 40W Fast Charging   
2    6 GB RAM, 128 GB inbuilt       5000 mAh Battery with Fast Charging   
3    6 GB RAM, 128 GB inbuilt   5000 mAh Battery with 30W Fast Charging   
4    6 GB RAM, 128 GB inbuilt   4700 mAh Battery with 66W Fast Charging   
..                        ...                                       ...   
95   8 GB RAM, 128 GB inbuilt       5000 mAh Battery with Fast Charging   
96    4 GB RAM, 64 GB inbuilt   5000 mAh Battery with 20W Fast Charging   
97  12 GB RAM, 256 GB inbuilt   4800 mAh Battery with 67W Fast Charging   
98   8 GB RAM, 256 GB inbuilt  4700 mAh Battery with 120W Fast Charging   
99   8 GB RAM, 128 GB inbuilt   4500 mAh Battery with 33W Fast Charging   

                                              display  \
0                    2.4 inches, 240 x 320 px Display   
1                 7.92 inches, 2088 x 2250 px Display   
2   6.55 inches, 1080 x 2400 px, 120 Hz Display wi...   
3   6.5 inches, 1080 x 2400 px, 120 Hz Display wit...   
4   6.44 inches, 1080 x 2404 px, 90 Hz Display wit...   
..                                                ...   
95  6.8 inches, 1080 x 2460 px Display with Punch ...   
96  6.53 inches, 720 x 1600 px, 90 Hz Display with...   
97  6.73 inches, 1440 x 3200 px, 120 Hz Display wi...   
98  6.78 inches, 1440 x 3200 px, 120 Hz Display wi...   
99  6.43 inches, 1080 x 2400 px Display with Punch...   

                                               camera  \
0                                  0.3 MP Rear Camera   
1           108 MP + 20 MP + 12 MP Triple Rear Camera   
2   50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
3   48 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
4   64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
..                                           

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  100 non-null    int64  
 1   model       100 non-null    object 
 2   price       100 non-null    int64  
 3   rating      78 non-null     float64
 4   sim         100 non-null    object 
 5   processor   100 non-null    object 
 6   ram         100 non-null    object 
 7   battery     100 non-null    object 
 8   display     100 non-null    object 
 9   camera      100 non-null    object 
 10  card        99 non-null     object 
 11  os          99 non-null     object 
dtypes: float64(1), int64(2), object(9)
memory usage: 9.5+ KB


### 2.Rating column has Some Mising values (nan values) , replace the nan values with the mean value of rating columns.

In [25]:
mean=int(df['rating'].mean())
mean

79

In [30]:
df['rating']=df['rating'].fillna(mean)
df['rating']

0     79.0
1     79.0
2     81.0
3     80.0
4     78.0
      ... 
95    79.0
96    75.0
97    79.0
98    88.0
99    82.0
Name: rating, Length: 100, dtype: float64

In [31]:
df

Unnamed: 0                 model   price  rating  \
0          917      Nokia 150 (2020)    2499    79.0   
1          462   Xiaomi Mi Mix Alpha  199990    79.0   
2            3  Motorola Moto G62 5G   14999    81.0   
3          739             Realme Q3   14999    80.0   
4          117        iQOO Z6 Pro 5G   21788    78.0   
..         ...                   ...     ...     ...   
95         682          LG V60 ThinQ   79990    79.0   
96         350     Motorola Moto G22    9499    75.0   
97         785    Xiaomi 12 Ultra 5G   75999    79.0   
98         594         iQOO 9 Pro 5G   59990    88.0   
99         297         OPPO F21s Pro   25895    82.0   

                                                  sim  \
0                                            Dual Sim   
1                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
2                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
3                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
4                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
..                                                ...   
95            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
96                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
97  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
98            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
99                 Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   

                                           processor  \
0                                            No Wifi   
1     Snapdragon 855+, Octa Core, 2.96 GHz Processor   
2      Snapdragon  695, Octa Core, 2.2 GHz Processor   
3    Snapdragon  750G , Octa Core, 2.2 GHz Processor   
4     Snapdragon  778G, Octa Core, 2.4 GHz Processor   
..                                               ...   
95    Snapdragon  865, Octa Core, 2.84 GHz Processor   
96          Helio  G37, Octa Core, 2.3 GHz Processor   
97  Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor   
98     Snapdragon 8 Gen1, Octa Core, 3 GHz Processor   
99     Snapdragon  695, Octa Core, 2.2 GHz Processor   

                          ram                                   battery  \
0                4 MB inbuilt                          1020 mAh Battery   
1   12 GB RAM, 512 GB inbuilt   4050 mAh Battery with 40W Fast Charging   
2    6 GB RAM, 128 GB inbuilt       5000 mAh Battery with Fast Charging   
3    6 GB RAM, 128 GB inbuilt   5000 mAh Battery with 30W Fast Charging   
4    6 GB RAM, 128 GB inbuilt   4700 mAh Battery with 66W Fast Charging   
..                        ...                                       ...   
95   8 GB RAM, 128 GB inbuilt       5000 mAh Battery with Fast Charging   
96    4 GB RAM, 64 GB inbuilt   5000 mAh Battery with 20W Fast Charging   
97  12 GB RAM, 256 GB inbuilt   4800 mAh Battery with 67W Fast Charging   
98   8 GB RAM, 256 GB inbuilt  4700 mAh Battery with 120W Fast Charging   
99   8 GB RAM, 128 GB inbuilt   4500 mAh Battery with 33W Fast Charging   

                                              display  \
0                    2.4 inches, 240 x 320 px Display   
1                 7.92 inches, 2088 x 2250 px Display   
2   6.55 inches, 1080 x 2400 px, 120 Hz Display wi...   
3   6.5 inches, 1080 x 2400 px, 120 Hz Display wit...   
4   6.44 inches, 1080 x 2404 px, 90 Hz Display wit...   
..                                                ...   
95  6.8 inches, 1080 x 2460 px Display with Punch ...   
96  6.53 inches, 720 x 1600 px, 90 Hz Display with...   
97  6.73 inches, 1440 x 3200 px, 120 Hz Display wi...   
98  6.78 inches, 1440 x 3200 px, 120 Hz Display wi...   
99  6.43 inches, 1080 x 2400 px Display with Punch...   

                                               camera  \
0                                  0.3 MP Rear Camera   
1           108 MP + 20 MP + 12 MP Triple Rear Camera   
2   50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
3   48 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
4   64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
..                                           

### 3.Extract the size of the display in inch and resolution size 1440 * 3216 px make the new columns of the screen size screen_size=4.1 inch,Xresolution=1440 and y_resolution=3216 and refresh rate=120hz from each row.

In [32]:
df.head(1)

Unnamed: 0             model  price  rating       sim processor  \
0         917  Nokia 150 (2020)   2499    79.0  Dual Sim   No Wifi   

            ram           battery                           display  \
0  4 MB inbuilt  1020 mAh Battery  2.4 inches, 240 x 320 px Display   

               camera                               card         os  
0  0.3 MP Rear Camera  Memory Card Supported, upto 32 GB  Bluetooth

In [64]:
df['screen_size']=df['display'].str.split(',').str.get(0).str.split().str.get(0).astype(float)
df['screen_size']

0     2.40
1     7.92
2     6.55
3     6.50
4     6.44
      ... 
95    6.80
96    6.53
97    6.73
98    6.78
99    6.43
Name: screen_size, Length: 100, dtype: float64

In [65]:
df['X-resolution']=df['display'].str.split(',').str.get(1).str.split('x').str.get(0).astype(float)
df['X-resolution']

0      240.0
1     2088.0
2     1080.0
3     1080.0
4     1080.0
       ...  
95    1080.0
96     720.0
97    1440.0
98    1440.0
99    1080.0
Name: X-resolution, Length: 100, dtype: float64

In [66]:
df['Y-resolution']=df['display'].str.split(',').str.get(1).str.split('x').str.get(1).str[:-1].astype(float)
df['Y-resolution']

0      320.0
1     2250.0
2     2400.0
3     2400.0
4     2404.0
       ...  
95    2460.0
96    1600.0
97    3200.0
98    3200.0
99    2400.0
Name: Y-resolution, Length: 100, dtype: float64

In [48]:
df['refresh rate']=df['display'].str.split(',').str.get(2).str.strip().str.split().str.get(0).fillna(0)

In [67]:
df

Unnamed: 0                 model   price  rating  \
0          917      Nokia 150 (2020)    2499    79.0   
1          462   Xiaomi Mi Mix Alpha  199990    79.0   
2            3  Motorola Moto G62 5G   14999    81.0   
3          739             Realme Q3   14999    80.0   
4          117        iQOO Z6 Pro 5G   21788    78.0   
..         ...                   ...     ...     ...   
95         682          LG V60 ThinQ   79990    79.0   
96         350     Motorola Moto G22    9499    75.0   
97         785    Xiaomi 12 Ultra 5G   75999    79.0   
98         594         iQOO 9 Pro 5G   59990    88.0   
99         297         OPPO F21s Pro   25895    82.0   

                                                  sim  \
0                                            Dual Sim   
1                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
2                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
3                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
4                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
..                                                ...   
95            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
96                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
97  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
98            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
99                 Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   

                                           processor  \
0                                            No Wifi   
1     Snapdragon 855+, Octa Core, 2.96 GHz Processor   
2      Snapdragon  695, Octa Core, 2.2 GHz Processor   
3    Snapdragon  750G , Octa Core, 2.2 GHz Processor   
4     Snapdragon  778G, Octa Core, 2.4 GHz Processor   
..                                               ...   
95    Snapdragon  865, Octa Core, 2.84 GHz Processor   
96          Helio  G37, Octa Core, 2.3 GHz Processor   
97  Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor   
98     Snapdragon 8 Gen1, Octa Core, 3 GHz Processor   
99     Snapdragon  695, Octa Core, 2.2 GHz Processor   

                          ram  battery  \
0                4 MB inbuilt     1020   
1   12 GB RAM, 512 GB inbuilt     4050   
2    6 GB RAM, 128 GB inbuilt     5000   
3    6 GB RAM, 128 GB inbuilt     5000   
4    6 GB RAM, 128 GB inbuilt     4700   
..                        ...      ...   
95   8 GB RAM, 128 GB inbuilt     5000   
96    4 GB RAM, 64 GB inbuilt     5000   
97  12 GB RAM, 256 GB inbuilt     4800   
98   8 GB RAM, 256 GB inbuilt     4700   
99   8 GB RAM, 128 GB inbuilt     4500   

                                              display  \
0                    2.4 inches, 240 x 320 px Display   
1                 7.92 inches, 2088 x 2250 px Display   
2   6.55 inches, 1080 x 2400 px, 120 Hz Display wi...   
3   6.5 inches, 1080 x 2400 px, 120 Hz Display wit...   
4   6.44 inches, 1080 x 2404 px, 90 Hz Display wit...   
..                                                ...   
95  6.8 inches, 1080 x 2460 px Display with Punch ...   
96  6.53 inches, 720 x 1600 px, 90 Hz Display with...   
97  6.73 inches, 1440 x 3200 px, 120 Hz Display wi...   
98  6.78 inches, 1440 x 3200 px, 120 Hz Display wi...   
99  6.43 inches, 1080 x 2400 px Display with Punch...   

                                               camera  \
0                                  0.3 MP Rear Camera   
1           108 MP + 20 MP + 12 MP Triple Rear Camera   
2   50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
3   48 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
4   64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
..                                                ...   
95                                       Dual Display   
96               50 MP Quad Rear & 16 MP Front Camera   
97  50 MP + 50 MP + 50 MP Triple Rear & 20 MP Fron...   
98  50 MP + 50 MP + 16 MP Triple Rear & 16 MP Fron...   
99  64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...   

                                                 card  \
0                   Memory Card Supported

### 4.Extract the battery in mAh from the battery column and convert into integer datatypes.

In [50]:
df.head(1)

Unnamed: 0             model  price  rating       sim processor  \
0         917  Nokia 150 (2020)   2499    79.0  Dual Sim   No Wifi   

            ram           battery                           display  \
0  4 MB inbuilt  1020 mAh Battery  2.4 inches, 240 x 320 px Display   

               camera                               card         os  \
0  0.3 MP Rear Camera  Memory Card Supported, upto 32 GB  Bluetooth   

  screen_size X-resolution Y-resolution refresh rate  
0  2.4 inches         240          320             0

In [55]:
df['battery']=df['battery'].str.split().str.get(0).astype(int)

In [56]:
df['battery']

0     1020
1     4050
2     5000
3     5000
4     4700
      ... 
95    5000
96    5000
97    4800
98    4700
99    4500
Name: battery, Length: 100, dtype: int64

In [57]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 16 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    100 non-null    int64  
 1   model         100 non-null    object 
 2   price         100 non-null    int64  
 3   rating        100 non-null    float64
 4   sim           100 non-null    object 
 5   processor     100 non-null    object 
 6   ram           100 non-null    object 
 7   battery       100 non-null    int64  
 8   display       100 non-null    object 
 9   camera        100 non-null    object 
 10  card          99 non-null     object 
 11  os            99 non-null     object 
 12  screen_size   100 non-null    object 
 13  X-resolution  99 non-null     object 
 14  Y-resolution  99 non-null     object 
 15  refresh rate  100 non-null    object 
dtypes: float64(1), int64(3), object(12)
memory usage: 12.6+ KB


### 5.Make a new column Pixel Per Inch(PPI) from the existing screen_size,X-resolution and Y-resolution.

In [69]:
import warnings
warnings.filterwarnings('ignore')
x=df['screen_size']
y=df['X-resolution'].fillna(0,inplace=True)
z=df['Y-resolution'].fillna(0,inplace=True)

In [70]:
df['PPI']=(((df['X-resolution'])**2+(df['Y-resolution']**2))**0.5)/df['screen_size']

In [71]:
df

Unnamed: 0                 model   price  rating  \
0          917      Nokia 150 (2020)    2499    79.0   
1          462   Xiaomi Mi Mix Alpha  199990    79.0   
2            3  Motorola Moto G62 5G   14999    81.0   
3          739             Realme Q3   14999    80.0   
4          117        iQOO Z6 Pro 5G   21788    78.0   
..         ...                   ...     ...     ...   
95         682          LG V60 ThinQ   79990    79.0   
96         350     Motorola Moto G22    9499    75.0   
97         785    Xiaomi 12 Ultra 5G   75999    79.0   
98         594         iQOO 9 Pro 5G   59990    88.0   
99         297         OPPO F21s Pro   25895    82.0   

                                                  sim  \
0                                            Dual Sim   
1                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
2                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
3                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
4                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
..                                                ...   
95            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
96                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
97  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
98            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
99                 Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   

                                           processor  \
0                                            No Wifi   
1     Snapdragon 855+, Octa Core, 2.96 GHz Processor   
2      Snapdragon  695, Octa Core, 2.2 GHz Processor   
3    Snapdragon  750G , Octa Core, 2.2 GHz Processor   
4     Snapdragon  778G, Octa Core, 2.4 GHz Processor   
..                                               ...   
95    Snapdragon  865, Octa Core, 2.84 GHz Processor   
96          Helio  G37, Octa Core, 2.3 GHz Processor   
97  Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor   
98     Snapdragon 8 Gen1, Octa Core, 3 GHz Processor   
99     Snapdragon  695, Octa Core, 2.2 GHz Processor   

                          ram  battery  \
0                4 MB inbuilt     1020   
1   12 GB RAM, 512 GB inbuilt     4050   
2    6 GB RAM, 128 GB inbuilt     5000   
3    6 GB RAM, 128 GB inbuilt     5000   
4    6 GB RAM, 128 GB inbuilt     4700   
..                        ...      ...   
95   8 GB RAM, 128 GB inbuilt     5000   
96    4 GB RAM, 64 GB inbuilt     5000   
97  12 GB RAM, 256 GB inbuilt     4800   
98   8 GB RAM, 256 GB inbuilt     4700   
99   8 GB RAM, 128 GB inbuilt     4500   

                                              display  \
0                    2.4 inches, 240 x 320 px Display   
1                 7.92 inches, 2088 x 2250 px Display   
2   6.55 inches, 1080 x 2400 px, 120 Hz Display wi...   
3   6.5 inches, 1080 x 2400 px, 120 Hz Display wit...   
4   6.44 inches, 1080 x 2404 px, 90 Hz Display wit...   
..                                                ...   
95  6.8 inches, 1080 x 2460 px Display with Punch ...   
96  6.53 inches, 720 x 1600 px, 90 Hz Display with...   
97  6.73 inches, 1440 x 3200 px, 120 Hz Display wi...   
98  6.78 inches, 1440 x 3200 px, 120 Hz Display wi...   
99  6.43 inches, 1080 x 2400 px Display with Punch...   

                                               camera  \
0                                  0.3 MP Rear Camera   
1           108 MP + 20 MP + 12 MP Triple Rear Camera   
2   50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
3   48 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
4   64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
..                                                ...   
95                                       Dual Display   
96               50 MP Quad Rear & 16 MP Front Camera   
97  50 MP + 50 MP + 50 MP Triple Rear & 20 MP Fron...   
98  50 MP + 50 MP + 16 MP Triple Rear & 16 MP Fron...   
99  64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...   

                                                 card  \
0                   Memory Card Supported

### 6.Extract the Ram and internal memory from the ram columns.

In [73]:
df.head(2)

Unnamed: 0                model   price  rating  \
0         917     Nokia 150 (2020)    2499    79.0   
1         462  Xiaomi Mi Mix Alpha  199990    79.0   

                                  sim  \
0                            Dual Sim   
1  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   

                                        processor                        ram  \
0                                         No Wifi               4 MB inbuilt   
1  Snapdragon 855+, Octa Core, 2.96 GHz Processor  12 GB RAM, 512 GB inbuilt   

   battery                              display  \
0     1020     2.4 inches, 240 x 320 px Display   
1     4050  7.92 inches, 2088 x 2250 px Display   

                                      camera  \
0                         0.3 MP Rear Camera   
1  108 MP + 20 MP + 12 MP Triple Rear Camera   

                                card           os  screen_size  X-resolution  \
0  Memory Card Supported, upto 32 GB    Bluetooth         2.40         240.0   
1          Memory Card Not Supported  Android v10         7.92        2088.0   

   Y-resolution refresh rate         PPI  
0         320.0            0  166.666667  
1        2250.0            0  387.571641

In [89]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 17 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    100 non-null    int64  
 1   model         100 non-null    object 
 2   price         100 non-null    int64  
 3   rating        100 non-null    float64
 4   sim           100 non-null    object 
 5   processor     100 non-null    object 
 6   ram           0 non-null      float64
 7   battery       100 non-null    int64  
 8   display       100 non-null    object 
 9   camera        100 non-null    object 
 10  card          99 non-null     object 
 11  os            99 non-null     object 
 12  screen_size   100 non-null    float64
 13  X-resolution  100 non-null    float64
 14  Y-resolution  100 non-null    float64
 15  refresh rate  100 non-null    object 
 16  PPI           100 non-null    float64
dtypes: float64(6), int64(3), object(8)
memory usage: 13.4+ KB


### 6.Extract the Ram and internal memory from the ram columns

In [5]:
df['Ram']=df['ram'].str.split(',').str[0][1:].str.split(" ").str[0]

In [6]:
df['Ram'].fillna('0GB',inplace=True)

C:\Users\DELL\AppData\Local\Temp\ipykernel_5532\1744292479.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Ram'].fillna('0GB',inplace=True)


In [7]:
df

Unnamed: 0                 model      price  rating  \
0          917      Nokia 150 (2020)     ₹2,499     NaN   
1          462   Xiaomi Mi Mix Alpha  ₹1,99,990     NaN   
2            3  Motorola Moto G62 5G    ₹14,999    81.0   
3          739             Realme Q3    ₹14,999    80.0   
4          117        iQOO Z6 Pro 5G    ₹21,788    78.0   
..         ...                   ...        ...     ...   
95         682          LG V60 ThinQ    ₹79,990     NaN   
96         350     Motorola Moto G22     ₹9,499    75.0   
97         785    Xiaomi 12 Ultra 5G    ₹75,999     NaN   
98         594         iQOO 9 Pro 5G    ₹59,990    88.0   
99         297         OPPO F21s Pro    ₹25,895    82.0   

                                                  sim  \
0                                            Dual Sim   
1                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
2                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
3                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
4                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
..                                                ...   
95            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
96                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
97  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
98            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
99                 Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   

                                           processor  \
0                                            No Wifi   
1     Snapdragon 855+, Octa Core, 2.96 GHz Processor   
2      Snapdragon  695, Octa Core, 2.2 GHz Processor   
3    Snapdragon  750G , Octa Core, 2.2 GHz Processor   
4     Snapdragon  778G, Octa Core, 2.4 GHz Processor   
..                                               ...   
95    Snapdragon  865, Octa Core, 2.84 GHz Processor   
96          Helio  G37, Octa Core, 2.3 GHz Processor   
97  Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor   
98     Snapdragon 8 Gen1, Octa Core, 3 GHz Processor   
99     Snapdragon  695, Octa Core, 2.2 GHz Processor   

                          ram                                   battery  \
0                4 MB inbuilt                          1020 mAh Battery   
1   12 GB RAM, 512 GB inbuilt   4050 mAh Battery with 40W Fast Charging   
2    6 GB RAM, 128 GB inbuilt       5000 mAh Battery with Fast Charging   
3    6 GB RAM, 128 GB inbuilt   5000 mAh Battery with 30W Fast Charging   
4    6 GB RAM, 128 GB inbuilt   4700 mAh Battery with 66W Fast Charging   
..                        ...                                       ...   
95   8 GB RAM, 128 GB inbuilt       5000 mAh Battery with Fast Charging   
96    4 GB RAM, 64 GB inbuilt   5000 mAh Battery with 20W Fast Charging   
97  12 GB RAM, 256 GB inbuilt   4800 mAh Battery with 67W Fast Charging   
98   8 GB RAM, 256 GB inbuilt  4700 mAh Battery with 120W Fast Charging   
99   8 GB RAM, 128 GB inbuilt   4500 mAh Battery with 33W Fast Charging   

                                              display  \
0                    2.4 inches, 240 x 320 px Display   
1                 7.92 inches, 2088 x 2250 px Display   
2   6.55 inches, 1080 x 2400 px, 120 Hz Display wi...   
3   6.5 inches, 1080 x 2400 px, 120 Hz Display wit...   
4   6.44 inches, 1080 x 2404 px, 90 Hz Display wit...   
..                                                ...   
95  6.8 inches, 1080 x 2460 px Display with Punch ...   
96  6.53 inches, 720 x 1600 px, 90 Hz Display with...   
97  6.73 inches, 1440 x 3200 px, 120 Hz Display wi...   
98  6.78 inches, 1440 x 3200 px, 120 Hz Display wi...   
99  6.43 inches, 1080 x 2400 px Display with Punch...   

                                               camera  \
0                                  0.3 MP Rear Camera   
1           108 MP + 20 MP + 12 MP Triple Rear Camera   
2   50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
3   48 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
4   64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
..       

In [8]:
df['Internal Memory']=df['ram'].str.split(',').str[1].str.split(' ').str[1]

In [9]:
df['Internal Memory'].fillna('4 MB',inplace=True)

C:\Users\DELL\AppData\Local\Temp\ipykernel_5532\1385856995.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Internal Memory'].fillna('4 MB',inplace=True)


In [10]:
df

Unnamed: 0                 model      price  rating  \
0          917      Nokia 150 (2020)     ₹2,499     NaN   
1          462   Xiaomi Mi Mix Alpha  ₹1,99,990     NaN   
2            3  Motorola Moto G62 5G    ₹14,999    81.0   
3          739             Realme Q3    ₹14,999    80.0   
4          117        iQOO Z6 Pro 5G    ₹21,788    78.0   
..         ...                   ...        ...     ...   
95         682          LG V60 ThinQ    ₹79,990     NaN   
96         350     Motorola Moto G22     ₹9,499    75.0   
97         785    Xiaomi 12 Ultra 5G    ₹75,999     NaN   
98         594         iQOO 9 Pro 5G    ₹59,990    88.0   
99         297         OPPO F21s Pro    ₹25,895    82.0   

                                                  sim  \
0                                            Dual Sim   
1                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
2                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
3                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
4                  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
..                                                ...   
95            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
96                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
97  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
98            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
99                 Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   

                                           processor  \
0                                            No Wifi   
1     Snapdragon 855+, Octa Core, 2.96 GHz Processor   
2      Snapdragon  695, Octa Core, 2.2 GHz Processor   
3    Snapdragon  750G , Octa Core, 2.2 GHz Processor   
4     Snapdragon  778G, Octa Core, 2.4 GHz Processor   
..                                               ...   
95    Snapdragon  865, Octa Core, 2.84 GHz Processor   
96          Helio  G37, Octa Core, 2.3 GHz Processor   
97  Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor   
98     Snapdragon 8 Gen1, Octa Core, 3 GHz Processor   
99     Snapdragon  695, Octa Core, 2.2 GHz Processor   

                          ram                                   battery  \
0                4 MB inbuilt                          1020 mAh Battery   
1   12 GB RAM, 512 GB inbuilt   4050 mAh Battery with 40W Fast Charging   
2    6 GB RAM, 128 GB inbuilt       5000 mAh Battery with Fast Charging   
3    6 GB RAM, 128 GB inbuilt   5000 mAh Battery with 30W Fast Charging   
4    6 GB RAM, 128 GB inbuilt   4700 mAh Battery with 66W Fast Charging   
..                        ...                                       ...   
95   8 GB RAM, 128 GB inbuilt       5000 mAh Battery with Fast Charging   
96    4 GB RAM, 64 GB inbuilt   5000 mAh Battery with 20W Fast Charging   
97  12 GB RAM, 256 GB inbuilt   4800 mAh Battery with 67W Fast Charging   
98   8 GB RAM, 256 GB inbuilt  4700 mAh Battery with 120W Fast Charging   
99   8 GB RAM, 128 GB inbuilt   4500 mAh Battery with 33W Fast Charging   

                                              display  \
0                    2.4 inches, 240 x 320 px Display   
1                 7.92 inches, 2088 x 2250 px Display   
2   6.55 inches, 1080 x 2400 px, 120 Hz Display wi...   
3   6.5 inches, 1080 x 2400 px, 120 Hz Display wit...   
4   6.44 inches, 1080 x 2404 px, 90 Hz Display wit...   
..                                                ...   
95  6.8 inches, 1080 x 2460 px Display with Punch ...   
96  6.53 inches, 720 x 1600 px, 90 Hz Display with...   
97  6.73 inches, 1440 x 3200 px, 120 Hz Display wi...   
98  6.78 inches, 1440 x 3200 px, 120 Hz Display wi...   
99  6.43 inches, 1080 x 2400 px Display with Punch...   

                                               camera  \
0                                  0.3 MP Rear Camera   
1           108 MP + 20 MP + 12 MP Triple Rear Camera   
2   50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
3   48 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
4   64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
..       